### DISCUSSION TITLE
Best strategy to store hierarchical JSON-like data

### DISCUSSION THREAD

**Discussion by johhnry** [prisma/prisma#15624] [Q&A] [answered]
Hi,

I am actually working on a project with Prisma and Postgres and I love it! ;)

However I am trying to store this simple data structure (TS interface example):

```ts
interface ClientInfo {
  id: string;
  hostname: string;
  os: {
    arch: string;
    platform: string;
    type: string;
    version: string;
    release: string;
  };
}
```

Since I want the `os` values to be separated/nested in the client data, I went with creating a second table for a one-to-one relation like this:

```prisma
model OS {
  id       Int     @id @default(autoincrement())
  arch     String
  platform String
  type     String
  version  String
  release  String
  Client   Client?
}

model Client {
  id       String @id
  hostname String

  os   OS       @relation(fields: [osId], references: [id])
  osId Int      @unique
}
```

Question: does it make sense to store the OS data in a separated table since the relation will only be between the Client and the OS? If the data doesn't change much is it really useful to have a relation between two tables?

I had issue when updating some client data with `upsert` and specially the nested relation with the `OS` model:

```ts
const { id, hostname, os } = clientInfo;
const record = { id, hostname };

await prisma.client.upsert({
  create: { ...record, os: { create: os } },
  update: { ...record, os: { connectOrCreate: { create: os, where: { id: /* ???? */ }} } },
  where: { id: clientInfo.id },
});
```

The thing is: if the client is not yet created, what should I put for the "where id: ????" close for the `OS` relation? I don't know in advance the id of the corresponding `OS` row.

This situation and complexity of updates / writes make me wonder if there's a simpler way to store that kind of hierarchical data.

- MongoDB should be more appropriate (composite types?) for that but I still need strong relations and queries that SQL provides.
- I could "flatten" the os data into the `Client` model by prefixing them with `os` like so:
  ```prisma
  model Client {
    osType String
    osArch String
    // ...
  }
  ```
- I can store data with the `Json` type in the schema but typing support is not optimal (https://github.com/prisma/prisma/issues/3219)

Thanks in advance for any insights / suggestions!

**Comment by austincrim** [comment, marked as answer]
Hey @johhnry 👋,

The three alternatives you laid out are the solutions I would look at as well. If I had to choose, I would go with the `Json` field and a little manual typing in your app code to improve the experience. That way, you don't have to restructure your data model or switch databases, and the nested data can change shape without much trouble.

Plus, when we ship typed `Json` fields, that drawback should disappear as well 🙂.

**Reply by johhnry** [reply]
Hi @austincrim,

Thanks for the answer! I might go with the `Json` field then, you convinced me.

Have a great day!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
